### ISSUE TITLE
BUG: regression in np.histogram

### ISSUE THREAD

**Issue by rgommers** [numpy/numpy#6120]
After gh-6100 was merged, 4 tests for ``scipy.stats.histogram`` (which uses ``np.histogram`` under the hood) have started failing:

    ======================================================================
    FAIL: test_increased_bins (test_stats.TestHistogram)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/rgommers/Code/scipy/scipy/stats/tests/test_stats.py", line 984, in test_increased_bins
        decimal=2)
      File "/home/rgommers/Code/numpy/numpy/testing/utils.py", line 510, in assert_almost_equal
        raise AssertionError(_build_err_msg())
    AssertionError: 
    Arrays are not almost equal to 2 decimals
     ACTUAL: 0
     DESIRED: 10

    ======================================================================
    FAIL: test_reduced_bins (test_stats.TestHistogram)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/rgommers/Code/scipy/scipy/stats/tests/test_stats.py", line 949, in test_reduced_bins
        decimal=2)
      File "/home/rgommers/Code/numpy/numpy/testing/utils.py", line 510, in assert_almost_equal
        raise AssertionError(_build_err_msg())
    AssertionError: 
    Arrays are not almost equal to 2 decimals
     ACTUAL: 0.07500000000000001
     DESIRED: -1.8749999999999996

    ======================================================================
    FAIL: test_simple (test_stats.TestHistogram)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/rgommers/Code/scipy/scipy/stats/tests/test_stats.py", line 876, in test_simple
        decimal=2)
      File "/home/rgommers/Code/numpy/numpy/testing/utils.py", line 886, in assert_array_almost_equal
        precision=decimal)
      File "/home/rgommers/Code/numpy/numpy/testing/utils.py", line 708, in assert_array_compare
        raise AssertionError(msg)
    AssertionError: 
    Arrays are not almost equal to 2 decimals

    (mismatch 20.0%)
     x: array([ 1.,  1.,  1.,  2.,  2.,  1.,  1.,  0.,  1.,  1.])
     y: array([ 1.,  1.,  1.,  2.,  1.,  2.,  1.,  0.,  1.,  1.])

    ======================================================================
    FAIL: test_weighting (test_stats.TestHistogram)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/rgommers/Code/scipy/scipy/stats/tests/test_stats.py", line 923, in test_weighting
        decimal=2)
      File "/home/rgommers/Code/numpy/numpy/testing/utils.py", line 886, in assert_array_almost_equal
        precision=decimal)
      File "/home/rgommers/Code/numpy/numpy/testing/utils.py", line 708, in assert_array_compare
        raise AssertionError(msg)
    AssertionError: 
    Arrays are not almost equal to 2 decimals

    (mismatch 20.0%)
     x: array([   4. ,    0. ,    4.5,   -0.9,    0. ,    0.3,  110.2,    0. ,
              0. ,   42. ])
     y: array([   4. ,    0. ,    4.5,   -0.9,    0. ,    0.3,    7. ,  103.2,
              0. ,   42. ])

    ----------------------------------------------------------------------

As this is a regression, it should be looked at for 1.10

**Comment by rgommers** [comment]
@astrofrog do you have time to look into this?

**Comment by jaimefrio** [comment]
The issue seems to be that the original array is modified by the call to `np.histogram`, and that's what's causing the SciPy failures. There was some extra logic added to avoid extra copies that we probably need to get rid off, starting [here](https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L231). I'll let @astrofrog take a look, in case he can find a less drastic solution, but if this turns into a 1.10 blocker we can fix it really quick.

**Comment by nayyarv** [comment]
Having recently merged this change to keep my PR up-to date, I noticed a couple of things

[This Line](https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L235) avoids the copy and the next 2 lines are inplace which is why the array passed in returns changed. Changing `copy = True` returns the array as is.

This fixes most of the scipy failures, though the test_weighting error still holds (but only for that single test case).

As far as I can tell, there is some weird float arithmetic going on here

```python
weights = np.array([1., 3., 4.5, 0.1, -1.0, 0.0, 0.3, 7.0, 103.2, 2, 40, 0, 0, 1])
low_values = np.array([0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.7, 0.8, 0.9, 1.1, 1.2], dtype=float)  # 11 values

#function calls blah blah
tmp_a = np.array(tmp_a, dtype=float, copy=True)
tmp_a -= mn
tmp_a *= norm

# check it's value *10 (handcoded, float)
tmp_a *10:     [   0.   10.   20.   30.   30.   40.   50.   60.   70.   90.  100.]
# ...
indices = tmp_a.astype(np.intp) # which bin each fits into

#float (handcoded)
indices:       [ 0  0  2  3  3  3  4  6  7  9 10]
#float32 (handcoded)
indices:       [ 0  1  2  3  3  4  5  6  7  9 10]
#float 128 (handcoded)
indices:       [0 0 2 2 2 3 4 6 7 9 9]

#float (np.arange used instead of handcoded)
indices:       [ 0  1  2  3  3  5  6  6  8  9 10]

#Finally note that
bins = array([ 0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,  1.2])) 
# i.e. the same as the values passed in 

# Expected indices by Scipy
indices: [0, 0, 2, 3, 3, 3, 5, 6, 6, 9, 9]
```
Changing the precision of the float of tmp_a (to say `np.float32, np.float64, np.float128`), changes results in this quite significantly.
Also, using `np.arange(0.2,1.3,0.1)` for `low_values` changes the results again.

My guess is that since the bins and the values are identical, you've simply got a matter of precision to deal with. A histogram worked out on paper (assuming left equality) would have indices be 1 per bin with 2 in the last bin. `indices = [0,1,2,3,4,5,6,7,8,9,9]`

This looks like some really odd floating point ambiguity that might be difficult to deal with consistently. Neither the old nor the new version is mathematically consistent in what it's trying to achieve, it appears that scipy simply had a unit test matching old behaviour rather than correctness.

**Comment by jaimefrio** [comment]
Thanks Varun, your analysis was very helpful!
 
For one thing, it seems to me that the scipy test is expecting wrong results to accommodate the inaccuracies existing in numpy's old implementation: the right answer for those indices should be `[ 0, 1, 2, 3, 3,4, 5, 6, 7, 9, 9]` I believe.

This happens to be the result you get if you cast the whole input array to `np.float32`, but I don't think that is going to turn into a fix for this issue. The fundamental problem is that the numbers in that test array do not have an exact floating point representation, and everything is downhill from there. If you replace the input array with those values multiplied by 10, everything works beautifully, because all the inputs are integers, which do have an exact floating point representation.

So we probably want to give it a little more thought, but my take is that the test is fundamentally flawed, so it is the test, not the function, that needs fixing.

**Comment by astrofrog** [comment]
@nayyarv - thanks for the analysis!

@jaimefrio - there is nevertheless a bug here which is that this line should be ``copy=True``:

https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L235

Pull request and regression test forthcoming!

**Comment by astrofrog** [comment]
@jaimefrio @rgommers @nayyarv - see https://github.com/numpy/numpy/pull/6126

**Comment by jaimefrio** [comment]
I've just merged #6126, but this is probably not going to fix all of the test cases in scipy, as discussed above.

**Comment by rgommers** [comment]
Thanks for the quick fix @jaimefrio @astrofrog @nayyarv! I'll have a look at what's left to fix in Scipy

**Comment by charris** [comment]
@rgommers How does this look? I'd like to branch but still see the failure with the problematical test case. I'm not convinced it is worth worrying about however.
```
FAIL: test_weighting (test_stats.TestHistogram)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/Workspace/scipy.git/build/testenv/lib64/python2.7/site-packages/scipy/stats/tests/test_stats.py", line 892, in test_weighting
    decimal=2)
  File "/home/charris/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 886, in assert_array_almost_equal
    precision=decimal)
  File "/home/charris/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 708, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 2 decimals

(mismatch 40.0%)
 x: array([   4. ,    0. ,    4.5,   -0.9,    0. ,    0.3,  110.2,    0. ,
          0. ,   42. ])
 y: array([   4. ,    0. ,    4.5,   -0.9,    0.3,    0. ,    7. ,  103.2,
          0. ,   42. ])
```

**Comment by rgommers** [comment]
@charris indeed not worth worrying about anymore. I guess we'll have to fix the scipy test and put it in a 0.16.1 to make the failure disappear.

**Comment by rgommers** [comment]
Closing this issue then. Thanks all for fixing the rest so quickly.
